src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs

```diff
@@ -414,21 +462,64 @@ internal ComponentInfo(Type interfaceType, string kind, Type argumentType, TlcMo
             ctor = null;
             create = null;
             requireEnvironment = false;
+            bool requireEnvironmentCtor = false;
+            bool requireEnvironmentCreate = false;
             var parmTypesWithEnv = Utils.Concat(new Type[1] { typeof(IHostEnvironment) }, parmTypes);
+
             if (Utils.Size(parmTypes) == 0 && (getter = FindInstanceGetter(instType, loaderType)) != null)
                 return true;
-            if (instType.IsAssignableFrom(loaderType) && (ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypes ?? Type.EmptyTypes, null)) != null)
-                return true;
-            if (instType.IsAssignableFrom(loaderType) && (ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypesWithEnv ?? Type.EmptyTypes, null)) != null)
+
+            // Find both 'ctor' and 'create' methods if available
+            if (instType.IsAssignableFrom(loaderType))
             {
-                requireEnvironment = true;
-                return true;
+                ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypes ?? Type.EmptyTypes, null);
+                if (ctor == null)
+                {
+                    ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypesWithEnv ?? Type.EmptyTypes, null);
+                    if (ctor != null)
+                        requireEnvironmentCtor = true;
+                }
             }
-            if ((create = FindCreateMethod(instType, loaderType, parmTypes ?? Type.EmptyTypes)) != null)
+
+            create = FindCreateMethod(instType, loaderType, parmTypes ?? Type.EmptyTypes);
+            if (create == null)
+            {
+                create = FindCreateMethod(instType, loaderType, parmTypesWithEnv ?? Type.EmptyTypes);
+                if (create != null)
+                    requireEnvironmentCreate = true;
+            }
+
+            if (ctor != null && create != null)
+            {
+                // If both 'ctor' and 'create' methods were found
+                // Choose the one that is 'more' public
+                // If they have the same visibility, then throw an exception, since this shouldn't happen.
+
+                if (ctor.Accessmodifier() == create.Accessmodifier())
+                {
```

Comment:
So I changed the nested if's for this other solution using the Accessmodifier() extension method. Although I think this one is more legible, I wouldn't be sure if it's worth it to create the extension method. So let me know your opinions, Thanks!

Reply:
To me this seems cleaner.

A couple of more ways you can decrease the amount of "if/else"s in the code:
1. you can put the assignment inside the if condition, like this:
```
if ((ctor = loaderType.GetConstructor(...)) == null)
```
2. If you throw or return inside the "if", then you don't need the "else":
```
if (ctor.Accessmodifier() == create.Accessmodifier())
    throw ...
if (ctor.Accessmodifier() > create.Accessmodifier())
{
    ...
    return true
}
if (ctor.Accessmodifier() < create.Accessmodifier())
...
etc.
```

---
In reply to: [347688260](https://github.com/dotnet/machinelearning/pull/4485#discussion_r347688260) [](ancestors = 347688260)